### PR TITLE
`<format>`: `std::format` should work for volatile arguments

### LINKED ISSUE #2427: `<format>`: `volatile integral<T>` compilation error
```cpp
#include <format>

int main()
{
    volatile int vol = 42;
    auto ret = std::format("{}", vol);
}
```
`error C2668: 'std::_Format_arg_traits<_Context>::_Phony_basic_format_arg_constructor': ambiguous call to overloaded function`

The same code compiles fine with **{fmt}**.
MSVS 17.0.4

Update:
It's rather `volatile integral types` fail to compile, because `volatile` `double` and `float` compile fine.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2427

### CODE REVIEW

**Comment by CaseyCarter** [comment]
I'm concerned this is not correctly handling reads from volatile objects - it's casting away volatile and using `memcpy` to copy byte-by-byte which is usually the wrong thing to do. This needs some thought.

**Comment by fsb4000** [comment]
I saw Charlie had changed  `_Phony_basic_format_arg_constructor` so it works with volatily types now, but `_Store_impl` still uses `memcpy` so I think this pull request is still needed.

**Comment by barcharcraz** [comment]
It's not clear what a pull request to fix this could do without kinda breaking the spirit of `volatile`

**Comment by fsb4000** [comment]
I agree that this is a little sloppy now because I started this PR before your PR was merged: https://github.com/microsoft/STL/pull/2323 and then I delete part of my PR.

https://stackoverflow.com/a/36729638

>memcpy may copy in any order, read parts of the source multiple times

So now, the PR tries to avoid read the volatile variable multiple times as @CaseyCarter noticed.

We just do a copy, and the compiler knows how to work with volatile variables so it will touch the variable only once and then we pass reference to copy and `memcpy` could read the copy multiple times in any parts order.

Also I decided to keep the regression test from issue 2427.

At least that's my intention in this PR.

**Review by barcharcraz — APPROVED** [review]
This now looks like an improvement over the current situation. Let's do it.

File: stl/inc/format [hunk 1]
```diff
@@ -1568,7 +1568,16 @@ private:
...
+        if constexpr (is_volatile_v<_Ty>) {
+            // _Erased_type is not a volatile type, so static_cast<_Erased_type> would cast away volatile
+            // _Store_impl uses memcpy for storing the value. We cannot use memcpy with volatile types.
+            // Because of the two reasons we have to make a temporary
+            static_assert(!is_volatile_v<_Erased_type>);
+            const _Erased_type _Temp = _Val;
+            _Store_impl<_Erased_type>(_Arg_index, _Arg_type, _Temp);
+        } else {
+            _Store_impl<_Erased_type>(_Arg_index, _Arg_type, static_cast<_Erased_type>(_Val));
+        }
```
Comment (StephanTLavavej) [inline stl/inc/format:1580]: While video reviewing this, @barcharcraz and @CaseyCarter helped me understand the history of this code and what the Standard says. After looking at this, we believe that `volatile int` is required to be ill-formed by the Standard, according to https://eel.is/c++draft/format.arg#4 . Specifically,

```cpp
typename Context::template formatter_type<remove_cvref_t<T>>()
  .format(declval<T&>(), declval<Context&>())
```

will pass `declval<volatile int&>()` to `formatter_type<int>`, and that will attempt to bind `.format()`'s parameter `const int&` to the lvalue argument of type `volatile int`, and that is not allowed.

@barcharcraz says that after #2323, the example in #2427 will be accepted (with a technically-undefined memcpying from `volatile`), but the Standard says we should reject this, so the code needs to be changed. I'm not immediately sure what the best change would be, but `static_assert(!is_volatile_v<remove_reference_t<_Ty>>)` (warning: not compiled, I just typed that out) might be a minimal fix. @barcharcraz and @CaseyCarter can help investigate.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Comment by fsb4000** [comment]
right solution: https://github.com/microsoft/STL/pull/2579

File: stl/inc/format (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/format:1580]: While it doesn't require library-defined `formatter` specializations to do so, I think the wording may allow a program-defined `formatter` specialization to accept `volatile` arguments. The stripping of volatile in [format.arg]/17.2 suggests to me that the drafters probably just didn't care enough to allow or forbid volatile.

**Pull request closed without merging**